pkg/datapath/linux/devices_controller.go

```diff
@@ -533,7 +534,7 @@ func (dc *devicesController) isSelectedDevice(d *tables.Device, routes statedb.T
 
 	log.Debug("Selected this device")
 
-	return true
+	return option.Config.AreDevicesRequired()
```

Comment:
CI failed because `isSelectedDevice()` should have returned `true` for eth0 when KPR enabled (which auto-enables NodePort), but that didn't happen.

https://github.com/cilium/cilium/blob/f439a14d8e007705edc1fc861bfa7d362c90ada3/daemon/cmd/kube_proxy_replacement.go#L74-L83

My guess is execution order: cilium-agent calls `isSelectedDevice()` ahead of `initKubeProxyReplacementOptions()`, so by the time executing `isSelectedDevice()`, option.EnableNodePort gets no chance to be turned on by `initKubeProxyReplacementOptions()`.

@joamaki I'd appreciate it if I can get some idea from you, once you are available someday this week.

Reply:
My plan is to over time remove `option.Config.Devices` and `AreDevicesRequired` completely and switch all modules to use the devices table directly. Then we would not have this central `AreDevicesRequired` check, but rather each feature (e.g. loader etc.) would itself figure out what to do with the information. Given that, I would put this `AreDevicesRequired` check into the "backwards compatible wrapper" (that wasn't): https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/devices.go#L37. E.g. return early:
```
  if len(userDevices) == 0 && !option.Config.AreDevicesRequired() {
    return nil, nil
  }
```

Reply:
> I would put this AreDevicesRequired check into the "backwards compatible wrapper"

I wonder how putting the check there solves the problem that the `deviceManager.Detect()` is called before the `initKubeProxyReplacement()`. The latter might set `option.Config.EnableNodePort` which influences the former.